Clean/Utils/Tactics/DecomposeProvableStruct.lean

```diff
@@ -0,0 +1,202 @@
+            match structArg with
+            | .fvar fvarId =>
+              -- Also continue searching in other arguments
+              let acc' ← args.foldlM (fun acc arg => go arg acc) acc
+              return fvarId :: acc'
+            | _ =>
+              -- Continue searching in all arguments
+              args.foldlM (fun acc arg => go arg acc) acc
+          else
+            -- Not enough arguments, continue searching
+            args.foldlM (fun acc arg => go arg acc) acc
+        else
+          -- Regular application, search in function and all arguments
+          let acc' ← go f acc
+          args.foldlM (fun acc arg => go arg acc) acc'
+      | _ =>
+        -- Regular application, search in function and all arguments
+        let acc' ← go f acc
+        args.foldlM (fun acc arg => go arg acc) acc'
+    | .lam _ _ body _ | .forallE _ _ body _ =>
+      -- For binders, check the body
+      go body acc
+    | .letE _ _ val body _ =>
+      -- For let expressions, check both value and body
+      let acc' ← go val acc
+      go body acc'
+    | .mdata _ expr => go expr acc
+    | _ => return acc
+
+  go e []
+
+/--
+  Find all variables that have ProvableStruct instances AND appear in field projections
+  in hypotheses or the goal (not just any variable with a ProvableStruct type)
+-/
+def findProvableStructVars : Lean.Elab.Tactic.TacticM (List Lean.FVarId) := do
+  withMainContext do
+    let ctx ← Lean.MonadLCtx.getLCtx
+    let mut result := []
+
+    -- Search for projections in the types of hypotheses
+    for localDecl in ctx do
+      if localDecl.isImplementationDetail then
+        continue
+
+      -- Search in the type of each hypothesis
+      let varsInType ← findStructVars localDecl.type
+      -- Filter to only include those with ProvableStruct instances
+      for fvarId in varsInType do
+        let type ← inferType (.fvar fvarId)
+        let type' ← withTransparency .reducible (whnf type)
+        let typeCtor := type'.getAppFn
+        match typeCtor with
+        | .const name _ =>
+          try
+            let instType ← mkAppM ``ProvableStruct #[.const name []]
+            if let .some _ ← trySynthInstance instType then
+              result := result ++ [fvarId]
+          catch _ => continue
+        | _ => continue
+
+    -- Also search for projections in the goal
+    let goal ← getMainGoal
+    let target ← goal.getType
+    let varsInGoal ← findStructVars target
+    -- Filter to only include those with ProvableStruct instances
+    for fvarId in varsInGoal do
+      let type ← inferType (.fvar fvarId)
+      let type' ← withTransparency .reducible (whnf type)
+      let typeCtor := type'.getAppFn
+      match typeCtor with
+      | .const name _ =>
+        try
+          let instType ← mkAppM ``ProvableStruct #[.const name []]
+          if let .some _ ← trySynthInstance instType then
+            result := result ++ [fvarId]
+        catch _ => continue
+      | _ => continue
```

Comment:
These lines are repeated for the goal and for each hypothesis. It's possible to refactor these.